R/buttons.R

```diff
@@ -0,0 +1,196 @@
+#' @param ... Named arguments become attributes to include on the `<button>`
+#'   element.
+#' @param auto_reset If `TRUE` (the default), automatically put the button back
+#'   in "ready" state after its click is handled by the server.
+#'
+#' @section Server value:
+#' An integer of class `"shinyActionButtonValue"`. This class differs from
+#' ordinary integers in that a value of 0 is considered "falsy".
+#' This implies two things:
+#'   * Event handlers (e.g., [shiny::observeEvent()], [shiny::eventReactive()]) won't execute on initial load.
+#'   * Input validation (e.g., [shiny::req()], [shiny::need()]) will fail on initial load.
+#'
+#' @examples
+#' if (interactive()) {
+#'   library(shiny)
+#'   library(bslib)
+#'
+#'   ui <- page_sidebar(
+#'     sidebar = sidebar(open = "always",
+#'       input_task_button("resample", "Resample"),
+#'     ),
+#'     verbatimTextOutput("summary")
+#'   )
+#'
+#'   server <- function(input, output, session) {
+#'     sample <- eventReactive(input$resample, ignoreNULL=FALSE, {
+#'       Sys.sleep(2)  # Make this artificially slow
+#'       rnorm(100)
+#'     })
+#'
+#'     output$summary <- renderPrint({
+#'       summary(sample())
+#'     })
+#'   }
+#'
+#'   shinyApp(ui, server)
+#' }
+#'
+#' @export
+input_task_button <- function(id, label, icon = NULL,
+  busy_label = "Processing...", busy_icon = icon_spinner(), ...,
+  type = "primary", auto_reset = TRUE) {
+
+  tags$button(
+    id = id,
+    class = if (!is.null(type)) paste0("btn btn-", type),
+    class = "bslib-task-button",
+    type = "button",
+    "data-auto-reset" = if (isTRUE(auto_reset)) NA else NULL,
+
+    component_dependencies(),
+
+    htmltools::tag("bslib-switch-inline",
+      list(
+        case = "ready",
+        span(slot = "ready",
+          icon, label
+        ),
+        span(slot = "busy",
+          busy_icon, busy_label
+        )
+      )
+    ),
+
+    ...
+  )
+}
+
+#' @rdname input_task_button
+#' @export
+icon_spinner <- function() {
+  fontawesome::fa_i("refresh", class = "fa-spin", "aria-hidden" = "true")
+}
+
+#' @param state If `"busy"`, put the button into busy/disabled state. If
+#'   `"ready"`, put the button into ready/enabled state.
+#' @param session The `session` object; using the default is recommended.
+#' @rdname input_task_button
+#' @export
+update_task_button <- function(id, state = c(NULL, "ready", "busy"), session = get_current_session()) {
+  force(id)
+
+  busy <- NULL
+
+  if (missing(state)) {
+  } else if (is.null(state)) {
+    # Do nothing
+  } else if (identical(state, "ready")) {
+    set_task_button_manual_reset(session, id, manual = FALSE)
+    busy <- FALSE
+  } else if (identical(state, "busy")) {
+    set_task_button_manual_reset(session, id, manual = TRUE)
+    busy <- TRUE
+  } else {
+    abort("Unexpected value for update_task_button's `state` argument; expected either \"ready\" or \"busy\"")
+  }
+
+  session$sendInputMessage(id, dropNulls(list(busy = busy)))
```

Comment:
Your version would work better for now, but I think the second you add the ability to change another property besides `"state"` (like `label`), you end up needing to rewrite. That's why I didn't return early in my version.

Reply:
Yeah, sorry, it would have been a lot more confusing to try to propose each separately. There's actually a third change I'm suggesting, too: don't error when `!state %in% c("ready", "busy")`. 

> Your version would work better for now, but I think the second you add the ability to change another property besides `"state"` (like `label`), you end up needing to rewrite.

I find the proposed code structure a lot easier to understand, which would also make it easier to rewrite safely in the future.

Reply:
After discussion: Don't error on unknown values. Simplify logic but don't return early.